atoma-state/src/state_manager.rs

```diff
@@ -241,35 +241,55 @@ impl AtomaState {
         &self,
         model: &str,
         free_units: i64,
-        owner: String,
         user_id: i64,
+        is_confidential: bool,
     ) -> Result<Vec<Stack>> {
-        // TODO: filter also by security level and other constraints
-        let tasks = sqlx::query(
-            "WITH selected_stack AS (
+        let mut query = String::from(
+            r#"
+            WITH selected_stack AS (
                 SELECT stacks.stack_small_id
                 FROM stacks
-                INNER JOIN tasks ON tasks.task_small_id = stacks.task_small_id
+                INNER JOIN tasks ON tasks.task_small_id = stacks.task_small_id"#,
+        );
+
+        if is_confidential {
+            query.push_str(r#"
+                INNER JOIN node_public_keys ON node_public_keys.node_small_id = stacks.node_small_id"#);
+        }
+
+        query.push_str(
+            r#"
                 WHERE tasks.model_name = $1
                 AND stacks.num_compute_units - stacks.already_computed_units >= $2
-                AND stacks.owner = $3
-                AND stacks.user_id = $4
+                AND stacks.user_id = $3"#,
+        );
+
+        if is_confidential {
+            query.push_str(
+                r#"
+                AND node_public_keys.is_valid = true"#,
+            );
+        }
+
+        query.push_str(
+            r#"
                 LIMIT 1
             )
             UPDATE stacks
             SET already_computed_units = already_computed_units + $2
             WHERE stack_small_id IN (SELECT stack_small_id FROM selected_stack)
-            RETURNING stacks.*",
-        )
-        .bind(model)
-        .bind(free_units)
-        .bind(owner)
-        .bind(user_id)
-        .fetch_all(&self.db)
-        .await?;
-        tasks
+            RETURNING stacks.*"#,
+        );
+
+        let stacks = sqlx::query(&query)
+            .bind(model)
+            .bind(free_units)
+            .bind(user_id)
+            .fetch_all(&self.db)
```

Comment:
I just noticed that we have LIMIT 1 so we can change this to fetch_optinal, but it doesn't have to be part of this PR, it's just an observation

Reply:
Because we can send only one to the node anyway. And you want to also update only one with the estimated compute units.